File: Dymarska2025_associations/generate_prompts.py
import pandas as pd
import json
import zipfile
from pathlib import Path

def generate_prompts():
    # Resolve from this script's location, not the working directory.
    base_dir = Path(__file__).resolve().parent
    processed_file = base_dir / "processed_data" / "exp1.csv"
    output_file = base_dir / "prompts.jsonl"

    if not processed_file.exists():
        print(f"Error: {processed_file} not found. Run preprocess_data.py first.")
        return

    # 1. Read the standardized data
    df = pd.read_csv(processed_file)

    # 2. Define instructions
    instructions = (
        "In this task, you will see a cue word, and you will be asked to type "
        "any associated words which come to mind, one by one.\n\n"
    )

    prompts = []

    # 3. Group by participant
    for p_id, group in df.groupby("participant_id"):
        group = group.sort_values("trial_id")
        
        prompt_text = instructions
        rt_list = []  # Initialize empty list for this specific participant

        # 4. Build trial-by-trial data
        for _, row in group.iterrows():
            trial_str = f"Trial {row['trial_id'] + 1}:\n"
            trial_str += f"  Stimulus: '{row['stimulus']}'\n"
            
            responses = []
            # We can collect responses AND RTs in the same loop for efficiency
            for i in range(1, 21):
                resp = row[f'response{i}']
                rt_val = row.get(f'first_key_RT{i}') # Using your specific RT column name

                if pd.notna(resp) and str(resp).strip() != "":
                    responses.append(f"<<{str(resp).strip()}>>")
                    
                    # Store RT if it exists and a response was given
                    if pd.notna(rt_val):
                        rt_list.append(float(rt_val))
            
            resp_string = ", ".join(responses)
            trial_line = f"{row['stimulus']}. You enter {resp_string}.\n"
            prompt_text += trial_line

        # 5. Create JSONL entry (Indented to be inside the participant loop, but outside the trial loop)
        entry = {
            "text": prompt_text.strip(),
            "experiment": "word_association_exp1",
            "participant_id": str(p_id),
            "rt": rt_list  # renamed from "rt_all": "rt" is the corpus-wide field name
        }
        prompts.append(entry)

    # 6. Write to JSONL
    with open(output_file, 'w', encoding='utf-8') as f:
        for p in prompts:
            f.write(json.dumps(p) + '\n')

    # Build the archive the repo tracks, so it is reproducible from this script
    # rather than zipped by hand -- which is where the stray __MACOSX/ entries in
    # several committed archives came from.
    zip_file = base_dir / "prompts.jsonl.zip"
    with zipfile.ZipFile(zip_file, "w", zipfile.ZIP_DEFLATED) as zf:
        zf.write(output_file, "prompts.jsonl")
    output_file.unlink()

    print(f"Successfully generated {len(prompts)} participant prompts in {zip_file}")

if __name__ == "__main__":
    generate_prompts()

File: petrenco2024_associations_images/generate_prompts.R
# =============================================================================
# generate_prompts.R
# -----------------------------------------------------------------------------
# Reads processed_data/exp1.csv and produces one prompt per participant,
# serialized as JSONL (one JSON object per line), then zipped.
#
# Experiment: Participants viewed images. In label conditions (positive/negative),
# a word appeared under the image 3 seconds after image onset. In the no-label
# condition, no word was shown. Participants generated 3 associations per image.

# Reads the tidy per-trial CSV written by preprocess_data.R and produces one
# narrative prompt per participant, serialized as JSONL (one JSON object per
# line) for downstream LLM evaluation.
#
# Pipeline:
#   processed_data/exp1.csv
#     |
#     v
#   prompts.jsonl
#     each line: {"text": <prompt>, "experiment": "petrenco2024_associations_images",
#                 "participant_id": <int>}
#
# Each participant's prompt consists of:
#   1. The task instructions shown to the human participant
#   2. A chronological replay of the participant's trials, one line per trial,
#      phrased as "What are the first 3 words that come to your mind when you think about this? ... You generate <<X>>,<<X>>,<<X>> ..."
#
# Usage:
#   Rscript generate_prompts.R
#
# Requires: readr, jsonlite  (install.packages(c("readr", "jsonlite")))
#
# Output: prompts.jsonl.zip
#   each line: {
#     "text": <chr>,
#     "experiment": <chr>,
#     "participant_id": <chr>,
#     "age": <dbl>,
#     "gender": <chr>,
#     "first_language": <chr>,
#     "list": <chr>
#   }
#
# Output: prompts.jsonl.zip
# =============================================================================

rm(list = ls())

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(purrr)
  library(jsonlite)
  library(zip)

# Resolve paths from this script's location so it runs from any working
# directory and always writes inside its own study folder.
.args <- commandArgs(trailingOnly = FALSE)
SCRIPT_DIR <- dirname(sub("^--file=", "", .args[grep("^--file=", .args)]))
if (length(SCRIPT_DIR) == 0 || !nzchar(SCRIPT_DIR)) SCRIPT_DIR <- getwd()

})

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------

EXPERIMENT_NAME <- "petrenco2024_associations_images" 

INSTRUCTIONS <- paste(
  "Experiment task",
  "You will first see <b>an image depicting a situation or a creature(s)",
  "In some cases you will then read words that will appear under the image.",
  "After you have looked at the image and read the word (if there is one), you will be asked to list the 3 first associations that come to your mind when you think about the creatures or situations depicted in the image.",
  "Try not to think too much and type in the first words that you thought of. Ensure they are different from each other.",
  "Once you have looked at the image, read the word, and proceeded to the next screen to type in the associations, you cannot go back to the previous screen with the image.",
  "Please pay attention and provide your associations immediately after viewing the image and reading the word (if there is one).",
  "Please pay attention during the task and try to come up with the associations right after you looked at the image and read the word (if there is one).",
  "You will not be able to type in the words that appeared under the images.",
  "If you do, you will be asked to match the requested format.",
  "Use letters only, no spaces, no punctuation, and no special characters",
  "Do not rush, the Continue button will be disabled for the first 5 seconds after the image presentation.",
  sep = "\n\n"
)

# -----------------------------------------------------------------------------
# Paths
# -----------------------------------------------------------------------------

input_file   <- file.path(SCRIPT_DIR, "processed_data", "exp1.csv")
output_jsonl <- file.path(SCRIPT_DIR, "prompts.jsonl")
output_zip   <- file.path(SCRIPT_DIR, "prompts.jsonl.zip")

# -----------------------------------------------------------------------------
# Read data
# -----------------------------------------------------------------------------

df <- read_csv(input_file, show_col_types = FALSE)

# Validate required columns
required_cols <- c(
  "participant_id", "trial_order", "condition",
  "stimulus1_image", "stimulus2_label", "stimulus2_present",
  "image_filename", "response1", "response2", "response3",
  "age", "gender", "first_language", "list"
)
missing_cols <- setdiff(required_cols, names(df))
if (length(missing_cols) > 0) {
  stop("Missing required columns in exp1.csv: ",
       paste(missing_cols, collapse = ", "))
}

df <- df %>% arrange(participant_id, trial_order)

# -----------------------------------------------------------------------------
# Build prompt for one participant
# -----------------------------------------------------------------------------

make_participant_prompt <- function(dat) {
  
  dat <- dat %>% arrange(trial_order)
  
  trial_lines <- map_chr(seq_len(nrow(dat)), function(i) {
    row <- dat[i, ]
    
    trial_num <- i  # 1-indexed
    
    # Stimulus line depends on condition
    if (!is.na(row$stimulus2_present) && row$stimulus2_present == "yes") {
      stimulus_line <- sprintf(
        "Trial %d: The image <%s> appears on the screen. Three seconds later, the word '%s' appears under the image.",
        trial_num, row$image_filename, row$stimulus2_label
      )
    } else {
      stimulus_line <- sprintf(
        "Trial %d: The image <%s> appears on the screen.",
        trial_num, row$image_filename
      )
    }
    
    # Response line: 3 associations marked with << >>.
    # Separated with ". " and not ", ": a ">>" followed directly by a comma is not
    # found by the training collator, and the failure is silent -- once a span
    # fails to close, the rest of the record collapses into one unmasked span.
    # See scripts/harmonization/bracket_safety.py.
    response_line <- sprintf(
      "You generate the following associations: <<%s>>. <<%s>>. <<%s>>.",
      row$response1, row$response2, row$response3
    )
    
    paste(stimulus_line, response_line, sep = "\n")
  })
  
  full_text <- paste0(
    INSTRUCTIONS, "\n\n",
    paste(trial_lines, collapse = "\n\n")
  )
  
  list(
    text           = full_text,
    experiment     = EXPERIMENT_NAME,
    participant_id = unique(dat$participant_id),
    age            = unique(dat$age),
    gender         = unique(dat$gender),
    first_language = unique(dat$first_language),
    list           = unique(dat$list)
  )
}

# -----------------------------------------------------------------------------
# Generate one record per participant
# -----------------------------------------------------------------------------

prompt_records <- df %>%
  group_by(participant_id) %>%
  group_split() %>%
  map(make_participant_prompt)

# -----------------------------------------------------------------------------
# Write JSONL
# -----------------------------------------------------------------------------

con <- file(output_jsonl, open = "w", encoding = "UTF-8")
for (rec in prompt_records) {
  writeLines(toJSON(rec, auto_unbox = TRUE), con = con)
}
close(con)

# -----------------------------------------------------------------------------
# Zip and clean up
# -----------------------------------------------------------------------------

if (file.exists(output_zip)) file.remove(output_zip)
zip::zipr(zipfile = output_zip, files = output_jsonl)
file.remove(output_jsonl)

# -----------------------------------------------------------------------------
# Summary and token limit check (~4 chars per token, 32K limit)
# -----------------------------------------------------------------------------

cat("Saved:", output_zip, "\n")
cat("Participants:", length(prompt_records), "\n")

char_limit <- 32000 * 4
texts      <- map_chr(prompt_records, "text")
over_limit <- which(nchar(texts) > char_limit)

if (length(over_limit) > 0) {
  warning("Participants exceeding ~32K token limit: ",
          paste(map_chr(prompt_records[over_limit], "participant_id"),
                collapse = ", "))
} else {
  cat("All prompts within 32K token limit.\n")
}

